src/Npgsql/TypeHandlers/CompositeHandlers/CompositeConstructorHandler.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Npgsql.PostgresTypes;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    class CompositeConstructorHandler<TComposite>
+    {
+        public PostgresType PostgresType { get; }
+        public ConstructorInfo ConstructorInfo { get; }
+        public CompositeParameterHandler[] Handlers { get; }
+
+        protected CompositeConstructorHandler(PostgresType postgresType, ConstructorInfo constructorInfo, CompositeParameterHandler[] handlers)
+        {
+            PostgresType = postgresType;
+            ConstructorInfo = constructorInfo;
+            Handlers = handlers;
+        }
+
+        public virtual async ValueTask<TComposite> Read(NpgsqlReadBuffer buffer, bool async)
+        {
+            await buffer.Ensure(sizeof(int), async);
+
+            var fieldCount = buffer.ReadInt32();
+            if (fieldCount != Handlers.Length)
+                throw new InvalidOperationException($"pg_attributes contains {Handlers.Length} fields for type {PostgresType.DisplayName}, but {fieldCount} fields were received.");
+
+            var args = new object?[Handlers.Length];
```

Comment:
This means that if the user has a type with over 8 members, and a constructor that corresponds to them, this will actually be less efficient than if there's no constructor and the members are writable - since we box everything.

Reply:
Yep, but on request we could increase this number. I think that 8 is a pretty good number to start with: not small, but not so big.

Reply:
At least worth keeping in mind that this PR actually regresses perf for some scenarios.